File: java/Day3.java
import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day3 {
    public static void main(String[] args) throws FileNotFoundException {
        long startTime = System.currentTimeMillis();
        
        Scanner in = new Scanner(new FileReader("input/day3.txt"));
        List<String> input = new ArrayList<>();
        while (in.hasNextLine()) input.add(in.nextLine());
        
        int total = part11(input);
        int remove = part1(input);
        System.out.println("total: " + total);
        System.out.println("removing: " + remove);
        System.out.println("Part I : " + (total - remove));
        //        System.out.println("Part II: " + part2(input));
        //        System.out.println("Execution time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
    
    private static int part11(List<String> input) {
        int row = input.size();
        int col = input.get(0).length();
        char[][] grid = new char[row][input.get(0).length()];
        for (int r = 0; r < row; r++) {
            grid[r] = input.get(r).toCharArray();
        }
        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                if (!Character.isDigit(grid[r][c]) && grid[r][c] != '.') {
                    grid[r][c] = '.';
                }
            }
        }
        
        Pattern pattern = Pattern.compile("\\d+");
        int total = 0;
        for (char[] chars : grid) {
            String line = new String(chars);
            Matcher matcher = pattern.matcher(line);
            while (matcher.find()) {
                int newNum = Integer.parseInt(matcher.group());
                //                System.out.println(newNum);
                total += newNum;
            }
        }
        return total;
    }
    
    private static int part1(List<String> input) {
        int row = input.size();
        int col = input.get(0).length();
        char[][] grid = new char[row][input.get(0).length()];
        for (int r = 0; r < row; r++) {
            grid[r] = input.get(r).toCharArray();
        }
        
        int[][] direct = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {1, -1}, {-1, 1}};
        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                //  && grid[r][c] != 'T'
                if (!Character.isDigit(grid[r][c]) && grid[r][c] != '.') {
                    //                    System.out.println(r + ", " + c + ": " + grid[r][c]);
                    for (int[] dr : direct) {
                        int rr = r + dr[0];
                        int cc = c + dr[1];
                        if (rr >= 0 && rr < row && cc >= 0 && cc < col && Character.isDigit(grid[rr][cc])) {
                            grid[rr][cc] = 'T';
                        }
                    }
                    grid[r][c] = 'T';
                }
            }
        }
        
        int res = 0;
        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                if (grid[r][c] == 'T') {
                    int idx = c + 1;
                    while (idx < col && Character.isDigit(grid[r][idx])) {
                        grid[r][idx] = 'T';
                        idx++;
                    }
                    idx = c - 1;
                    while (idx >= 0 && Character.isDigit(grid[r][idx])) {
                        grid[r][idx] = 'T';
                        idx--;
                    }
                }
                if (Character.isDigit(grid[r][c])) {
                    StringBuilder part = new StringBuilder();
                    int idx = c;
                    while (idx < col) {
                        if (grid[r][idx] == '.') {
                            c = idx;
                            System.out.println(Integer.parseInt(part.toString()));
                            res += Integer.parseInt(part.toString());
                            break;
                        } else if (grid[r][idx] == 'T') {
                            break;
                        } else {
                            part.append(grid[r][idx]);
                            //                            grid[r][idx] = 'N';
                            idx++;
                        }
                    }
                }
            }
        }
        
        printer(grid);
        
        return res; // not 515900, 508206
    }
    
    private static void printer(char[][] grid) {
        for (char[] chars : grid) {
            System.out.println(Arrays.toString(chars));
        }
    }
    
    private static int part2(List<String> input) {
        return -1;
    }
}

File: rust/src/bin/day3.rs
use lazy_static::lazy_static;

lazy_static!(
    static ref FILE: String = std::fs::read_to_string("../input/day3.txt").unwrap();
);
const DIRECT: [(i32, i32); 8] = [(0, 1), (1, 0), (0, -1), (-1, 0), (1, 1), (-1, -1), (-1, 1), (1, -1)];

fn main() {
    println!("{}{}", "Expect 514969, Result: ", part1(&FILE));
    println!("{}{}", "Expect 78915902, Result: ", part2(&FILE));
}

fn part1(input: &str) -> i32 {
    let mut res = 0;
    let mut grid: Vec<Vec<char>> = input.lines().map(|line| line.chars().collect()).collect();

    for r in 0..grid.len() {
        for c in 0..grid[r].len() {
            if !grid[r][c].is_numeric() && grid[r][c] != '.' {
                for &(dr, dc) in &DIRECT {
                    let nr = r as i32 + dr;
                    let nc = c as i32 + dc;
                    if nr >= 0 && nr < grid.len() as i32 && nc >= 0 && nc < grid[r].len() as i32 && grid[nr as usize][nc as usize].is_digit(10) {
                        res += get_nums(&mut grid, nr as usize, nc as usize);
                    }
                }
            }
        }
    }
    res
}

fn get_nums(grid: &mut Vec<Vec<char>>, r: usize, c: usize) -> i32 {
    let mut c = c;
    let col = grid[r].len();
    while c > 0 && grid[r][c - 1].is_digit(10) {
        c -= 1;
    }
    let mut res = String::new();
    while c < col && grid[r][c].is_digit(10) {
        res.push(grid[r][c]);
        grid[r][c] = 'N';
        c += 1;
    }
    res.parse::<i32>().unwrap()
}

fn part2(input: &str) -> i32 {
    let mut res = 0;
    let mut grid: Vec<Vec<char>> = input.lines().map(|line| line.chars().collect()).collect();

    for r in 0..grid.len() {
        for c in 0..grid[r].len() {
            if grid[r][c] == '*' {
                let nums: Vec<i32> = DIRECT.iter().filter_map(|&(dr, dc)| {
                    let (nr, nc) = (r as i32 + dr, c as i32 + dc);
                    if nr >= 0 && nr < grid.len() as i32 && nc >= 0 && nc < grid[r].len() as i32 && grid[nr as usize][nc as usize].is_digit(10) {
                        Some(get_nums(&mut grid, nr as usize, nc as usize))
                    } else {
                        None
                    }
                }).collect();
                if nums.len() == 2 {
                    res += nums[0] * nums[1];
                }
            }
        }
    }
    res
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn day3_part1() {
        assert_eq!(514969, part1(&FILE))
    }

    #[test]
    fn day3_part2() {
        assert_eq!(78915902, part2(&FILE))
    }
}
